libsolidity/analysis/NameAndTypeResolver.cpp

```diff
@@ -476,7 +476,15 @@ bool DeclarationRegistrationHelper::registerDeclaration(
 	Declaration const* shadowedDeclaration = nullptr;
 	if (_warnOnShadow && !name.empty() && _container.enclosingContainer())
 		for (auto const* decl: _container.enclosingContainer()->resolveName(name, true, true))
-			shadowedDeclaration = decl;
+			// Do not warn about functions shadowing a contract.
```

Comment:
Can you explain the reason for this?

Reply:
I would say the warning is correct, since it does shadow `A`. Note the following does not work because the function `A` shadows the contract `A`:
```
contract A { function f() public {} }
contract B is A {
  function A() public pure returns (uint8) {}
  function g() public {
    A.f();
  }
}
```